### PR TITLE
feat: Create Conversation form & template

### PR DESCRIPTION
**Pull request by hobbyhacker0**
## Description

Form & Template for creating new Conversation

Depends on the Licence / Application query in [VOL-4737](https://dvsa.atlassian.net/browse/VOL-4737)

Related issue: [VOL-4681](https://dvsa.atlassian.net/browse/VOL-4681)

## Before submitting (or marking as "ready for review")

- [x] Does the pull request title follow the [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/) specification?
- [x] Have you performed a self-review of the code
- [ ] Have you have added tests that prove the fix or feature is effective and working
- [ ] Did you make sure to update any documentation relating to this change?

### CODE REVIEW

File: module/Olcs/src/Controller/Messages/LicenceNewConversationController.php [hunk 1]
```diff
@@ -31,4 +40,38 @@ public function getLeftView()
 
         return $view;
     }
+
+    public function alterFormForAdd($form){
```
Comment (wadedvsa) [inline module/Olcs/src/Controller/Messages/LicenceNewConversationController.php:44]: S: New line for `}` and whitespace only on 45?

File: module/Olcs/src/Controller/Messages/LicenceNewConversationController.php [hunk 2]
```diff
@@ -31,4 +40,38 @@ public function getLeftView()
...
+
+        $appLicNoSelect = $form->get('fields')->get('appLicNo');
+
+        /**
+         * var Common\Service\Cqrs\Response $data
```
Comment (wadedvsa) [inline module/Olcs/src/Controller/Messages/LicenceNewConversationController.php:49]: S: `@var`?

File: module/Olcs/src/Controller/Messages/LicenceNewConversationController.php [hunk 3]
```diff
@@ -31,4 +40,38 @@ public function getLeftView()
...
+         */
+        $data = $this->handleQuery(
+            ByLicenceToOrganisation::create(['licence' => $this->params()->fromRoute('licence')])
+        );
+
+        $applicationLicenceArray = json_decode($data->getHttpResponse()->getBody(), true);
+
+        $options = array();
```
Comment (wadedvsa) [inline module/Olcs/src/Controller/Messages/LicenceNewConversationController.php:57]: S: `[]` over `array()` for consistency with the code below.

File: module/Olcs/src/Controller/Messages/LicenceNewConversationController.php (hunk 2 above)
Comment (hobbyhacker0) [reply module/Olcs/src/Controller/Messages/LicenceNewConversationController.php:49]: Ooh good spot

**Review by wadedvsa — DISMISSED** [review]

File: module/Olcs/src/Controller/IndexController.php [hunk 4]
```diff
@@ -263,6 +263,11 @@ public function entityListAction()
                 ->setIrhpPermitStock($value);
                 $results = ['' => 'Please select'] + $srv->fetchListOptions();
                 break;
+            case '':
```
Comment (sdh100shaun) [inline module/Olcs/src/Controller/IndexController.php:266]: Can we review what type will be here

Comment (hobbyhacker0) [reply module/Olcs/src/Controller/IndexController.php]: The '' case isn't needed, just left over from a previous experimentation, have checked and cleaned up now.

There may be a case for making the Licence / Application query behaviour more similar to the other cases populated by queries (so the code in the controller's alterFormForAdd() would be refactored to align with the getListDataUser() and getListDataDocTemplates())

As a Random Act of Kindness we could remove the repeated fetchListOptions() by replacing with a single array and addition (where needed) before the format & return

**Review by wadedvsa — APPROVED** [review]

File: module/Olcs/src/Controller/AbstractNewConversationController.php [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+        if ($this->params()->fromRoute('licence')){
+            $licence_id = $this->params()->fromRoute('licence');
+            $data = $this->handleQuery(
+                ByLicenceToOrganisation::create(['licence' => $licence_id])
+            );
+        } elseif ($this->params()->fromRoute('application')) {
+            $application_id = $this->params()->fromRoute('application');
+            $data = $this->handleQuery(
+                ByApplicationToOrganisation::create(['application' => $application_id])
```
Comment (wadedvsa) [inline module/Olcs/src/Controller/AbstractNewConversationController.php:56]: A: snake_case -> camelCase

File: module/Olcs/src/Controller/Messages/AbstractNewConversationController.php [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+        if ($this->params()->fromRoute('licence')){
+            $licence_id = $this->params()->fromRoute('licence');
+            $data = $this->handleQuery(
+                ByLicenceToOrganisation::create(['licence' => $licence_id])
+            );
+        } elseif ($this->params()->fromRoute('application')) {
+            $application_id = $this->params()->fromRoute('application');
+            $data = $this->handleQuery(
+                ByApplicationToOrganisation::create(['application' => $application_id])
```
Comment (wadedvsa) [inline module/Olcs/src/Controller/Messages/AbstractNewConversationController.php:56]: A: snake_case -> camelCase

File: module/Olcs/src/Controller/Messages/AbstractNewConversationController.php [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+            );
+            $routeMatch = $this->getEvent()->getRouteMatch();
+            $routeMatch->setParam('', 'new_value');
+        } else {
+            throw new \RuntimeException('Error: licence or application required');
```
Comment (wadedvsa) [inline module/Olcs/src/Controller/Messages/AbstractNewConversationController.php:61]: S: Import

File: module/Olcs/src/Controller/AbstractNewConversationController.php [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+            );
+            $routeMatch = $this->getEvent()->getRouteMatch();
+            $routeMatch->setParam('', 'new_value');
+        } else {
+            throw new \RuntimeException('Error: licence or application required');
```
Comment (wadedvsa) [inline module/Olcs/src/Controller/AbstractNewConversationController.php:61]: S: Import

File: module/Olcs/src/Controller/Messages/ApplicationNewConversationController.php [hunk 9]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+namespace Olcs\Controller\Messages;
+
+// use \Olcs\Controller\Messages\AbstractNewConversationController;
```
Comment (wadedvsa) [inline module/Olcs/src/Controller/Messages/ApplicationNewConversationController.php:5]: S: Remove commented line

File: module/Olcs/src/Controller/Messages/LicenceNewConversationController.php [hunk 10]
```diff
@@ -2,33 +2,10 @@
 
 namespace Olcs\Controller\Messages;
 
-use Laminas\View\Model\ViewModel;
-use Olcs\Controller\AbstractInternalController;
+// use \Olcs\Controller\Messages\AbstractNewConversationController;
```
Comment (wadedvsa) [inline module/Olcs/src/Controller/Messages/LicenceNewConversationController.php:5]: S: Remove commented line

**Review by wadedvsa — APPROVED** [review]

**Pull request merged**